### ISSUE TITLE
timeout in Package Deployer task

### ISSUE THREAD

**Issue by lheng18** [WaelHamze/dyn365-ce-vsts-tasks#71]
@WaelHamze Got the following Timeout issue in Package Deployer task. It failed when deleting the holding solution, eventually worked but it shows failed in the log. Is it possible you can set the timeout longer when call CRM API? Thanks!

Error	: The request channel timed out while waiting for a reply after 00:02:00. Increase the timeout value passed to the call to Request or increase the SendTimeout value on the Binding. The time allotted to this operation may have been a portion of a longer timeout.
 Error	: The HTTP request to 'https://myapps-stg.crm.dynamics.com/XRMServices/2011/Organization.svc' has exceeded the allotted timeout of 00:02:00. The time allotted to this operation may have been a portion of a longer timeout.
2019-04-25T15:40:39.3372782Z Stack Trace	: at System.ServiceModel.Channels.HttpChannelUtilities.ProcessGetResponseWebException(WebException webException, HttpWebRequest request, HttpAbortReason abortReason)
2019-04-25T15:40:39.3372974Z    at System.ServiceModel.Channels.HttpChannelFactory`1.HttpRequestChannel.HttpChannelRequest.WaitForReply(TimeSpan timeout)
2019-04-25T15:40:39.3373161Z    at System.ServiceModel.Channels.RequestChannel.Request(Message message, TimeSpan timeout)

**Comment by lheng18** [comment]
@WaelHamze maybe you can 
[change the timeout property ](https://community.dynamics.com/365/b/sachinbansalmicrosoft/archive/2018/10/29/the-request-channel-timed-out-while-waiting-for-a-reply-after-00-01-59-9396415-increase-the-timeout-value-passed-to-the-call-to-request-or-increase-the-sendtimeout-value-on-the-binding)

Fix:

To resolve this issue, we increased Timeout property. OrganizationServiceProxy class has Timeout property which can be set to increase default value of 2 min as below: For more details please refer here.


((OrganizationServiceProxy)(orgService)).Timeout = new TimeSpan(0, 10, 0);

Add this line of code after orgService object is created above to set Timeout to 10 minutes.

In case you are using CrmConnection class, then you can directly set Timeout property using CrmConnection class object as below:


CrmConnection conn = new CrmConnection(strUrl);

conn.Timeout = new TimeSpan(0, 10, 0);

Thanks!

**Comment by aelsmore** [comment]
@lheng18 - which timeout property did you increase? Was it the property in the task, or in the connection string?

@WaelHamze - I have the same issue too. Always on the Uninstall step of an Import/Upgrade of a managed solution.

> TimeoutException - **DeleteAndPromote** : Delete and Promote Solution to New version(**DeleteAndPromoteSolutionRequest**) |=> The request channel timed out while waiting for a reply after 00:02:00

Is there any chance this timeout can be increased, or honour the one specified on the task?

**Comment by aelsmore** [comment]
@WaelHamze - Would adding -MaxCrmConnectionTimeOutMinutes $Timeout parameter to the Get-CrmConnection call be a potential resolution here? I'm happy to create a pull request if that will speed up a turnaround?

$CRMConn = Get-CrmConnection -ConnectionString $CrmConnectionString -MaxCrmConnectionTimeOutMinutes $Timeout -Verbose

**Comment by lheng18** [comment]
@aelsmore Yes we had the same issue. Not sure if your fix would work. Let's wait for @WaelHamze to answer. Thanks!

**Comment by WaelHamze** [comment]
@lheng18 @aelsmore I have added another timeout attribute on the task, this one will be passed into the connection, the other one is passed to the package.

This should be available in the next release hopefully next week

**Comment by aelsmore** [comment]
Hi @WaelHamze, any update on when the next release will drop?

**Comment by aelsmore** [comment]
Not sure whether this helps, but the issue doesn't happen in 9.* of the package deployer task

**Comment by lheng18** [comment]
It does happen in 9.*. @WaelHamze , I got one more problem. When we used it to upgrade a solution, we got "Solution dependencies exist, cannot uninstall.Dependency" besides the error we mentioned above 2019-06-03T18:14:54.2635419Z PackageDeployer Information: 8 : ****** PACKAGE DEPLOYMENT PROCESS COMPLETED. Result:FAILED  Duration:00:17:44.8661364 ******
2019-06-03T18:14:54.4573855Z ##[error]Updating  MySolution version 1.9.19126.2 to version 2.9.19154.1: Deleting Holding solution.
2019-06-03T18:14:54.4649440Z ##[error]Failed to delete solution MySolution , version: 2.9.19154.1. See the log file.
2019-06-03T18:14:54.6981927Z ##[error]Import Failed

**Comment by lheng18** [comment]
@WaelHamze , manually importing the zip file generated from the Drop asset that the above release pipeline was tried to import works fine. So it is the issue with the Deployer task, is it the Microsoft Packager deployer SDK tool issue or Build Task issue?

**Comment by HIREN-AITC** [comment]
I am getting CRM Timeout error while importing managed solution using MSCRM Package Deployer. I understand that in the latest release of 365 Build Tools 9.8.4, it includes parameter CRMConnectionTimeout to increase CRM connection timeout. But I don't see that parameter in my task in my pipeline. I have the latest version of 365 Build tools in my DEVOPS. I am not able to understand how do I pass this parameter to the task?
![image](https://user-images.githubusercontent.com/42240205/59087768-f5038080-8905-11e9-83fc-fb5140695f44.png)
![image](https://user-images.githubusercontent.com/42240205/59087793-051b6000-8906-11e9-9d1c-cf5c03794765.png)
![image](https://user-images.githubusercontent.com/42240205/59087866-372cc200-8906-11e9-940b-4aa16288eba8.png)

**Comment by lheng18** [comment]
@WaelHamze Has the new connection timeout attribute feature been released?

**Comment by WaelHamze** [comment]
@lheng18 not yet will be released shortly.

**Comment by lheng18** [comment]
@WaelHamze Thanks for your reply. Do you know when it will be released?

**Comment by WaelHamze** [comment]
Released

**Comment by lheng18** [comment]
@WaelHamze, thanks for publish the new release. I still encounter the timeout issue. I always set Packager Deployer Timeout to be 2:45:00, 2 hours and 45 minutes, also I just set the CRM Connection Timeout to be 120 minutes. Normally the deployment only takes 30 minutes. But it took 2 hours not completed until it is expired. 


 Error	: The HTTP request to 'https://my-test.crm.dynamics.com/XRMServices/2011/Organization.svc' has exceeded the allotted timeout of **00:02:00**. The time allotted to this operation may have been a portion of a longer timeout.
2019-06-23T21:14:44.8373248Z Stack Trace	: at System.ServiceModel.Channels.HttpChannelUtilities.ProcessGetResponseWebException(WebException webException, HttpWebRequest request, HttpAbortReason abortReason)
2019-06-23T21:14:44.8373696Z    at System.ServiceModel.Channels.HttpChannelFactory`1.HttpRequestChannel.HttpChannelRequest.WaitForReply(TimeSpan timeout)
2019-06-23T21:14:44.8374014Z    at System.ServiceModel.Channels.RequestChannel.Request(Message message, TimeSpan timeout)

I noticed,it still shows  allotted timeout of **00:02:00**. How do I leverage the new CRM Connection settings?

@aelsmore @HIREN-AITC Do you still encounter the same issue?
Thanks!

**Comment by HIREN-AITC** [comment]
@Iheng18 No I am not facing this issue now but Deleting holding solution fails even after waiting for 30 minutes. 
You can check if the latest version of the extension is deployed or not on your debops.

**Comment by lheng18** [comment]
@HIREN-AITC @WaelHamze yes, i still got the same unable to delete the solution due timeout issue. I am using the latest version.

**Comment by lheng18** [comment]
I still have this issue. WaelHamze, can you reopen the issue?

**Comment by rojorioko** [comment]
Hi @WaelHamze, I am also experiencing this issue while deploying the package.

![image](https://user-images.githubusercontent.com/54129932/70770286-ba443e80-1da7-11ea-9d75-1d97270eabe2.png)

**Comment by KimBon** [comment]
Hi @WaelHamze , @rojorioko

I have tested "import-crmpackage"  directly from PowerShell, without using any of the scripts within your toolset. I have tested with following configurations (both timeouts 120 mins):
-  First test, with same configuration as this project
  - Microsoft.Xrm.Tooling.CrmConnector.Powershell: 9.0.0.9073
  - Microsoft.Xrm.Tooling.PackageDeployment.Powershell: 3.0.0.975
- Second test, newest packages on PowerShell gallery
  - Microsoft.Xrm.Tooling.CrmConnector.Powershell: 3.3.0.873
https://www.powershellgallery.com/packages/Microsoft.Xrm.Tooling.CrmConnector.PowerShell
  - Microsoft.Xrm.Tooling.PackageDeployment.Powershell: 3.3.0.874
https://www.powershellgallery.com/packages/Microsoft.Xrm.Tooling.PackageDeployment.Powershell

The first test resulted in the same issue as we have in the pipelines, and does not directly seem related to any of the timeouts already implemented:

![afbeelding](https://user-images.githubusercontent.com/19949103/70810317-d8b83300-1dc3-11ea-8313-90548553700f.png)

In the end, the apply solution upgrade happened but the result is not captured correctly in PowerShell due to this issue.

The second test used the latest versions of both PowerShell libraries, and seem handle the Apply Solution Upgrade just fine:

![afbeelding](https://user-images.githubusercontent.com/19949103/70810696-862b4680-1dc4-11ea-8fc3-e6ea69d57f8d.png)

As you can see, Deleting the holding solution worked and the rest of the execution proceeded.

Is there any reason not to upgrade the libraries for this?

Separate question: would you know if its possible to stage for upgrade via package deployer but not forcing the delete of the holding and have another task do this?

Thanks for considering an update on these libraries.

Kind regards

Kim

**Comment by WaelHamze** [comment]
@KimBon @HIREN-AITC @lheng18 thanks for the investigation.

I have just updated the tool to the latest connectors and PS libraries.

I normally use the import/apply tasks to import big solutions as I get full control over the timeouts etc.. using these and never have any issues with the solutions I worked with so far.

I believe the Package Deployer Task offers all the available parameters you get with the PS module. Let me know if you have any suggestions for this.
